### PR TITLE
Add Micrometer support for Task applications

### LINKED ISSUE #608: Add Micrometer support for Task applications
Similar to what we have done in Spring Batch (see: [BATCH-2774](https://jira.spring.io/browse/BATCH-2774)), we could add support to track Task execution statistics in Micrometer.

Maybe:
- Task Duration
- Result
- ..

### PR DESCRIPTION
**Pull request by tzolov**
- New dependencies
   - micrometer-core
   - spring-boot-starter-actuator - required for the custom micrometer tags conf.
   - java-cfenv - required by the CF tags configuration.
 - Add TaskMetrics configuration that installs two Meters (Timer and LongTaskTimer) that produce the following time-series:
   - spring.cloud.task - shows task duration (updated only after the task has completed).
   - spring.cloud.task.active - provides run time information about the non-completed tasks.
   Both meters are started at TaskExecutionListener#onTaskStartup() and ended at onTaskEnd() or onTaskFailure().
 - Add common Task tags assigned to all time-series produced by the Task (not only the s.c.task and s.c.task.active). This allows to analyze CPU, Memory and other measurements grouped by task/execution ids.
 - Add Cloud Found tags - Activated only for CF target platform.

 Resolves #608

### CODE REVIEW

File: spring-cloud-task-core/src/test/java/org/springframework/cloud/task/micrometer/SpringCloudTaskMicrometerCommonTagsConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+			assertThat(meter.getId().getTag("task.parent.execution.id"))
+					.isEqualTo("unknown");
+			assertThat(meter.getId().getTag("task.external.execution.id"))
+					.isEqualTo("unknown");
+		}
+
+	}
+
+	@TestPropertySource(properties = { "spring.cloud.task.name=myTask",
+			"spring.cloud.task.executionid=666",
```
Comment (mminella) [inline spring-cloud-task-core/src/test/java/org/springframework/cloud/task/micrometer/SpringCloudTaskMicrometerCommonTagsConfigurationTest.java]: Given the connotations with 666, we may want to choose another value here.

File: spring-cloud-task-core/src/test/java/org/springframework/cloud/task/micrometer/TaskMetricsConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Christian Tzolov
+ */
+@RunWith(Enclosed.class)
+public class TaskMetricsConfigurationTest {
+
+	@TestPropertySource(properties = { "spring.cloud.task.name=myTask",
+			"spring.cloud.task.executionid=666" })
```
Comment (mminella) [inline spring-cloud-task-core/src/test/java/org/springframework/cloud/task/micrometer/TaskMetricsConfigurationTest.java]: See previous note about 666.

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/TaskMetricsConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * @author Christian Tzolov
+ */
+@Configuration
+public class TaskMetricsConfiguration {
+
+	private static final String METRICS_PREFIX = "spring.cloud.";
```
[the remark this answers is not in the document]
Comment (tzolov) [reply spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/TaskMetricsConfiguration.java]: This prefix is useless. We should use the full meter names as constants. Like this:
```
public static final String SPRING_CLOUD_TASK_METER = "spring.cloud.task";
public static final String SPRING_CLOUD_TASK_ACTIVE_METER = "spring.cloud.task.active";
```

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/TaskMetricsConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(getTaskName(), getExecutionId());
+		}
+
+	}
+
+	public static class MicrometerTaskExecutionListener
```
[the remark this answers is not in the document]
Comment (tzolov) [reply spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/TaskMetricsConfiguration.java]: I've been considering the same. Using the listener seems cleaner but could have side effects. Including the listeners time durations in the task's measurement is one of the issue. Another more severe IMO is that if one listener before the metrics listener fail then the metrics listener is never called.

File: spring-cloud-task-core/pom.xml [hunk 5]
```diff
@@ -106,6 +106,19 @@
...
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>io.micrometer</groupId>
+			<artifactId>micrometer-core</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-actuator</artifactId>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (tzolov) [reply spring-cloud-task-core/pom.xml]: This is hard one as we discussed already. It would require further considerations.

**Comment by tzolov** [comment]
@mminella, fyi the  33b9eda commit should address most of the comments except the actuator dependency.

**Comment by mminella** [comment]
I've reviewed Actuator and how `MeterRegistryCustomizer` implementations are used.  Really all that happens is that there is a `BeanPostProcessor` that loops over any of the customizer implementations and calls configure anytime a bean of type `MeterRegistry` is created (see [here](https://github.com/spring-projects/spring-boot/blob/aef92b9295f62d008faa9ab79905a474bf3496f3/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/MeterRegistryPostProcessor.java#L62)).  Adding the Actuator library just to handle that dozen lines of code feels like a lot unless I'm missing something.  I really don't see how bringing Actuator in cuts down on the lines of code we need to maintain.  Instead of implementing a `MeterRegistryCustomizer`, we implement a `BeanPostProcessor`.  Thoughts?

**Comment by tzolov** [comment]
The `MeterRegistryCustomizer` is only one of the places where the Actuator is being used for the Tasks-micrometer integration. Here are two other usages i can think of:
- The Spring Cloud Task App Starters need the actuator to manage the multiple meter-repos on the classpath (e.g. influx, prometheus, datadog, prometheus.rsocket ...). The `SpringCloudTaskMicrometerEnvironmentPostProcessor`  leverages the `management.metrics.export.[influx|datadog|prometheus|prometheus.rsocket].enabled` properties to manage this.
- The DataFlow server uses the `management.metrics.export.<meterregistry>.xxx` to configure the desired meter registry.  For example the SCDF influxdb configuration for task apps would look something like this: 
```
- spring.cloud.dataflow.applicationProperties.task.management.metrics.export.influx.enabled=true
- spring.cloud.dataflow.applicationProperties.task.management.metrics.export.influx.db=myinfluxdb
- spring.cloud.dataflow.applicationProperties.task.management.metrics.export.influx.uri=http://influxdb:8086
```
For example here the `management.metrics.export.influx.db` pre-creates the `myinfluxdb` database if it doesn't exist and this functionality exists only inside the actuator.  In fact most [meter-repos management in boot ](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#production-ready-metrics-getting-started) appears to be implemented inside the actuator.

While we might be able to get away by forking the `MeterRegistryCustomizer` we would still need the actuator for the Task App Starters and DataFlow Task management. If later is true, then it would be cleaner to move `CloudFoundryMicrometerTagsConfiguration` and `SpringCloudTaskMicrometerCommonTagsConfiguration` to the Task App Starter Core (where the actuator dependency is present) instead of reimplementing `MeterRegistryCustomizer` inside SCT?

**Comment by mminella** [comment]
We need to look at these PRs in isolation because they will be used that way.  Not all SCT users use app starters or SCDF.  Because of that, I still think that my position stands for this SCT PR.  The addition of actuator for the app starters sounds like it makes sense in that use case (multiple registries on the classpath) and the required integration support for SCDF.

**Comment by tzolov** [comment]
@mminella,  this is your call :)  
Committed a cannibalized, stripped down version of the  MeterRegistryCustomizer infrastructure.  Hopefully it will have no side effects with future extensions (such as auto-configuration ordering). 
My  opinion remains that those custom tags should go to the Task App starter.

**Comment by mminella** [comment]
Rebased, polished, and merged.

**Pull request closed without merging**